File: code/dynamic-conditional.kt
// The dynamic type is not supported in code targeting the JVM
// https://kotlinlang.org/docs/reference/dynamic-type.html

// Use .let and forget weird helper methods

val data = DataPoint(1, 22.1, DataPoint(2, 22.8))

val result = data.child?.child?.
            let { toFahrenheit(it.celsius) } ?: Double.MIN_VALUE

File: code/dynamic-conditional.cs
// Dynamic helper method that will return boolean and set output
bool GetValue(dynamic? input, out dynamic? output)
{
    output = input;
    return output != null;
}

var data = new DataPoint 
{ 
    Id = 1, 
    Celsius = 22.1, 
    Child = new DataPoint { Id = 2, Celsius = 22.8 }
};

var result = GetValue(data.Child?.Child?.Celsius, out var output)
                ? ToFahrenheit(output) 
                : double.MinValue;

var text = GetValue("My text", out var output) ? output : "Not set";
// "My text"
var text = GetValue(null, out var output) ? output : "Not set";
// "Not set"
